cognite/client/_api/assets.py

```diff
@@ -641,234 +706,169 @@ def _get_children(self, assets: List) -> List:
         return children
 
 
-class _AssetsFailedToPost:
-    def __init__(self, exc: Exception, assets: Sequence[Asset]):
-        self.exc = exc
-        self.assets = assets
+class _AssetHierarchyCreator:
+    def __init__(self, hierarchy: AssetHierarchyValidator, assets_api: AssetsAPI) -> None:
+        hierarchy.is_valid(on_error="raise")
+        self.hierarchy = hierarchy
+        self.assets_api = assets_api
+        self.create_limit = assets_api._CREATE_LIMIT
+        self.failed: List[Asset] = []
+        self.unknown: List[Asset] = []
+        self.exception: Union[CogniteAPIError, CogniteDuplicatedError, None] = None
+
+    def create(self, upsert: bool, upsert_mode: UpsertOptions) -> AssetList:
+        insert_dct = self.hierarchy.groupby_parent_xid()
+        insert_fn = functools.partial(self._insert, upsert=upsert, upsert_mode=upsert_mode)
+        created_assets = AssetList([], cognite_client=self.assets_api._cognite_client)
+
+        # Kick things off with all...:
+        # 1. Root assets
+        # 2. Assets linking parent by ID
+        # 3. Orphans assets (if `hierarchy.ignore_orphans` is True)
+        to_create = insert_dct.pop(None)
+
+        # Optimization: Child assets can be created alongside parents -if- part of the same request.
+        self._extend_unblocked_from_subtree(to_create, insert_dct)
```

Comment:
Can you do a comparison of performance on the old and new implementations? I think losing the concurrency will have significant impact.

Reply:
Should only be slightly worse in contrived scenarios: the typical hierarchy is growing geometrically with depth, thus any slowness in the start (to get the roots and first few levels unblocked) is the only time when the old double-concurrent solution is faster, as we get to the "lower levels", the number of assets >> number-of-workers and the waiting time is purely backend dependent. Tests also show this.

To combat some of that "initial slowness", `_extend_unblocked_from_subtree` will make sure to add at least one full request with roots and its children (>= 1000 assets), since these can be created simultaneously.

```
# Test 1:
- 262 142 assets
- hierarchy: networkx.balanced_tree(2, 17)
- 2: branching factor
- 17: depth

# Test 2:
- 50 100 assets
- hierarchy: 100 levels with 100 assets,
  each new level picks parents randomly from previous

# Test 1, new code:
- insert: 68.80 sec
# Test 1, old code:
- insert: 66.51 sec

# Test 2, old code
- insert: 24.89 sec
- insert: 17.92 sec
# Test 2, new code:
- insert: 29.41 sec
- insert: 26.99 sec
```

Reply:
<= 1000 *

Reply:
Nice!

Reply:
I was ofc not happy with the new code potentially being slower in synthetic scenarios ( 😂 ) so I've rewritten it 🤓 It is now has the exact same performance with the potential to be faster (queued requests now use `PriorityThreadPoolExecutor` with the size of the subtree as the priority) 🚀 .